### PR TITLE
Ultra Honk arithmetic and grand product relations

### PR DESCRIPTION
**Pull request by ledwards2225**
# Description

Add the arithmetic and permutation grand product relations for Ultra Honk. (Total of 4 relations) (Part of issue https://github.com/AztecProtocol/barretenberg/issues/342)

New relations (all names are fully flexible at this point):
- `UltraArithmeticRelation`: Primary arithmetic relation
- `UltraArithmeticRelationSecondary`: Secondary arithmetic relation for non-native field op applications
- `UltraGrandProductInitializationRelation`: Equivalent to Standard version
- `UltraGrandProductComputationRelation`: Equivalent to Standard version

This work also adds tests for all of these relations, following the format of the existing relations tests. I've reorganized all relations tests into two suites: 
- `RelationConsistency`: Check that arithmetic in the relation is equivalent to some simplified implementation of the same arithmetic. (Useful for when we start optimizing).
- `RelationCorrectness`: Check that the constraint encoded in each relation is actually satisfied for the polynomials produced by the composer for a real circuit.

Other changes:
- Some templatizing of test fixtures in `RelationConsistency` to account for different flavor-esque parameters
- I've narrowed the scope of the Plonk-Honk composer consistency tests to account for some correct and expected differences between the polynomials produced by the respective composers. They're still useful for checking the witnesses and table polys but are close to the end of their useful life.

Notes:
- We may be able to combine the permutation grand product relations for Standard/Ultra via templating later on if desired. This is tricky now because of the rigid way we access polynomials but should be made easier by Cody's flavor updates. On the other hand, a `Relation` is a pretty simple object and it might not hurt to have some code duplication there for simplicity.
- All of the naming in this PR is up for debate. I anticipate making some changes and am open to suggestions

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR specifications in `/markdown/specs` have been updated.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] I've updated any terraform that needs updating (e.g. environment variables) for deployment.
- [ ] The branch has been rebased against the head of its merge target.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.

### CODE REVIEW

File: cpp/src/barretenberg/honk/composer/composer_helper/ultra_honk_composer_helper.cpp [hunk 1]
```diff
@@ -201,11 +201,15 @@ std::shared_ptr<plonk::proving_key> UltraHonkComposerHelper<CircuitConstructor>:
 
     construct_lagrange_selector_forms(circuit_constructor, circuit_proving_key.get());
 
-    enforce_nonzero_polynomial_selectors(circuit_constructor, circuit_proving_key.get());
+    // TODO(#217)(luke): Naively enforcing non-zero selectors for Honk will result in some relations not being
+    // satisfied.
+    // enforce_nonzero_polynomial_selectors(circuit_constructor, circuit_proving_key.get());
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/composer/composer_helper/ultra_honk_composer_helper.cpp:206]: If we do this, we'll have to add the zero knowledge relation before we can create a valid proof, since selectors could easily be zero.

File: cpp/src/barretenberg/honk/composer/ultra_honk_composer.test.cpp [hunk 2]
```diff
@@ -39,14 +39,19 @@ std::vector<uint32_t> add_variables(auto& composer, std::vector<fr> variables)
...
 {
     // Check that all lagrange polys agree
     auto& honk_store = honk_prover.key->polynomial_store;
     auto& plonk_store = plonk_prover.key->polynomial_store;
     for (auto& entry : honk_store) {
         std::string key = entry.first;
-        if (plonk_store.contains(key)) {
+        bool is_sorted_table = (key.find("s_") != std::string::npos);
+        bool is_table = (key.find("table_value_") != std::string::npos);
+        if (plonk_store.contains(key) && (is_sorted_table || is_table)) {
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/composer/ultra_honk_composer.test.cpp:54]: Why are you not checking wires any more?

File: cpp/src/barretenberg/honk/flavor/flavor.hpp [hunk 3]
```diff
@@ -187,4 +187,58 @@ struct StandardHonk {
...
 };
+
+struct UltraArithmetization {
+    /**
+     * @brief All of the multivariate polynomials used by the Ultra Honk Prover.
+     * @details The polynomials are broken into three categories: precomputed, witness, and shifted.
+     * This separation must be maintained to allow for programmatic access, but the ordering of the
+     * polynomials can be permuted within each category if necessary. Polynomials can also be added
+     * or removed (assuming consistency with the prover algorithm) but the constants describing the
+     * number of poynomials in each category must be manually updated.
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/flavor/flavor.hpp]: Typo "poynomials"

File: cpp/src/barretenberg/honk/flavor/flavor.hpp [hunk 4]
```diff
@@ -187,4 +187,58 @@ struct StandardHonk {
...
+        W_O,
+        W_4,
+        S_1,
+        S_2,
+        S_3,
+        S_4,
+        Z_PERM,
+        Z_LOOKUP,
+        /* --- SHIFTED POLYNOMIALS --- */
+        W_1_SHIFT,
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/flavor/flavor.hpp:236]: W_L_SHIFT?

**Review by Rumata888 — APPROVED** [review]
I am a bit worried about the verify_consistency function, otherwise everything looks great

File: cpp/src/barretenberg/honk/flavor/flavor.hpp (hunk 4 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/flavor/flavor.hpp:236]: `W_1_SHIFT` was intentional since I prefer the numerical index but you're right I'm not being consistent here. I'll make them all numeric in an immediate follow on, but regardless, all of this Flavor stuff I'm adding will go away entirely with Cody's incoming flavor work.

File: cpp/src/barretenberg/honk/flavor/flavor.hpp (hunk 3 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/flavor/flavor.hpp]: Fixed, thanks

File: cpp/src/barretenberg/honk/composer/ultra_honk_composer.test.cpp (hunk 2 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/composer/ultra_honk_composer.test.cpp:54]: The wires are checked just below this. It has to be done in a separate loop since the wires are owned directly by the prover in Honk but they are still stored in the proving_key for Plonk. I've added a comment to clarify.

File: cpp/src/barretenberg/honk/composer/composer_helper/ultra_honk_composer_helper.cpp (hunk 1 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/composer/composer_helper/ultra_honk_composer_helper.cpp:206]: Its a very good point. We need to make sure our tests are not passing trivially with zero selectors until we decide how to proceed. Maybe we should prioritize issue #217 in the near future.
